src/functions.php

```diff
@@ -241,3 +242,41 @@ function recursive_copy($element) {
 
     return clone $element;
 }
+
+/**
+ * Creates a type map to apply to a field type
+ *
+ * This is used in the Aggregate, Distinct, and FindAndModify operations to
+ * apply the root-level type map to the document that will be returned. It also
+ * replaces the root type with object for consistency within these operations
+ *
+ * An existing type map for the given field path will not be overwritten
+ *
+ * @internal
+ * @param array  $typeMap   The existing typeMap
+ * @param string $fieldPath The field path to apply the root type to
+ * @return array
+ */
+function create_field_path_type_map(array $typeMap, $fieldPath)
+{
+    if (isset($typeMap['root']) && ! isset($typeMap['fieldPaths'][$fieldPath])) {
```

Comment:
Should we skip this just because the user has specified both a custom `root` type _and_ happens to have specified our internal field path in their own `fieldPaths` option?

For instance, if we're using this with FindAndModify and the user specified a custom type for the `value` field path, I imagine we should just move that to the `value.value` field path. And independent of that, whatever the user specifies for `root` can always can always be assigned to `$typeMap['fieldPaths'][$fieldPath]` at the end (as you do later in the function).

I wonder if it'd make sense to just handle these independently.

Reply:
You are correct. The helper function now always nests all pre-existing field paths to prefix them with the given field path. It also applies an `object` type for the `root` key, and an `array` type for the `array` key if the given field path ends in `.$`. This was added because the Aggregate, Distinct, and MapReduce operations expect an array of results instead of an iterator. We may always optimise this further down the road by returning an iterator directly instead of wrapping the results after the call. For now, this works just fine though.